File: app/schemas/journal.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class JournalEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trade_id: int | None
    tradingsymbol: str
    entry_type: str
    content: str
    tags: str
    strategy: str
    outcome: str
    emotional_state: str
    created_at: datetime
    updated_at: datetime


class JournalEntryCreateRequest(BaseModel):
    tradingsymbol: str
    entry_type: str  # pre_trade / post_trade / note
    content: str = ""
    tags: str = ""
    strategy: str = ""
    outcome: str = ""  # win / loss / breakeven
    emotional_state: str = ""  # confident / anxious / fomo / revenge / neutral
    trade_id: int | None = None


class JournalEntryUpdateRequest(BaseModel):
    tradingsymbol: str | None = None
    entry_type: str | None = None
    content: str | None = None
    tags: str | None = None
    strategy: str | None = None
    outcome: str | None = None
    emotional_state: str | None = None
    trade_id: int | None = None


class StrategyStats(BaseModel):
    strategy: str
    count: int
    wins: int
    losses: int
    win_rate: float


class TagStats(BaseModel):
    tag: str
    count: int


class JournalAnalytics(BaseModel):
    total_entries: int
    win_rate: float
    total_wins: int
    total_losses: int
    total_breakeven: int
    by_strategy: list[StrategyStats]
    by_tag: list[TagStats]

File: frontend/src/types/journal.ts
export interface JournalEntry {
  id: number;
  trade_id: number | null;
  tradingsymbol: string;
  entry_type: string;
  content: string;
  tags: string;
  strategy: string;
  outcome: string;
  emotional_state: string;
  created_at: string;
  updated_at: string;
}

export interface JournalCreateRequest {
  tradingsymbol: string;
  entry_type: string;
  content?: string;
  tags?: string;
  strategy?: string;
  outcome?: string;
  emotional_state?: string;
  trade_id?: number | null;
}

export interface JournalUpdateRequest {
  tradingsymbol?: string;
  entry_type?: string;
  content?: string;
  tags?: string;
  strategy?: string;
  outcome?: string;
  emotional_state?: string;
  trade_id?: number | null;
}

export interface StrategyStats {
  strategy: string;
  count: number;
  wins: number;
  losses: number;
  win_rate: number;
}

export interface TagStats {
  tag: string;
  count: number;
}

export interface JournalAnalytics {
  total_entries: number;
  win_rate: number;
  total_wins: number;
  total_losses: number;
  total_breakeven: number;
  by_strategy: StrategyStats[];
  by_tag: TagStats[];
}

export interface BehaviorFlag {
  id: number;
  flag_type: string;
  severity: string;
  description: string;
  trade_id: number | null;
  is_acknowledged: boolean;
  created_at: string;
}

export interface BehaviorDetectionResponse {
  flags: BehaviorFlag[];
  summary: string;
}

export interface BehaviorSummary {
  total: number;
  by_severity: Record<string, number>;
  unacknowledged: number;
  recent_flags: BehaviorFlag[];
}
